frontends/p4/frontend.cpp

```diff
@@ -227,30 +230,37 @@ const IR::P4Program *FrontEnd::run(const CompilerOptions &options, const IR::P4P
         new RemoveAllUnusedDeclarations(&refMap),
         new RemoveRedundantParsers(&refMap, &typeMap),
         new ClearTypeMap(&typeMap),
-        evaluator,
-        new Inline(&refMap, &typeMap, evaluator, options.optimizeParserInlining),
-        new InlineActions(&refMap, &typeMap),
-        new LocalizeAllActions(&refMap),
-        new UniqueNames(&refMap),
-        new UniqueParameters(&refMap, &typeMap),
-        // Must be done before inlining functions, to allow
-        // function calls used as action arguments to be inlined
-        // in the proper place.
-        new RemoveActionParameters(&refMap, &typeMap),
-        new InlineFunctions(&refMap, &typeMap),
-        new SetHeaders(&refMap, &typeMap),
-        // Check for constants only after inlining
-        new CheckConstants(&refMap, &typeMap),
-        new SimplifyControlFlow(&refMap, &typeMap),
-        new RemoveParserControlFlow(&refMap, &typeMap),  // more ifs may have been added to parsers
-        new UniqueNames(&refMap),                        // needed again after inlining
-        new MoveDeclarations(),                          // needed again after inlining
-        new SimplifyDefUse(&refMap, &typeMap),
-        new RemoveAllUnusedDeclarations(&refMap),
-        new SimplifyControlFlow(&refMap, &typeMap),
+        evaluator
+    });
+    if (options.optimizationLevel > 0)
+        passes.addPasses({
+            new Inline(&refMap, &typeMap, evaluator, options.optimizeParserInlining),
+            new InlineActions(&refMap, &typeMap),
+            new LocalizeAllActions(&refMap),
+            new UniqueNames(&refMap),
+            new UniqueParameters(&refMap, &typeMap),
+            // Must be done before inlining functions, to allow
+            // function calls used as action arguments to be inlined
+            // in the proper place.
+            new RemoveActionParameters(&refMap, &typeMap),
+            new InlineFunctions(&refMap, &typeMap),
```

Comment:
Some later stages that don't know about functions or action calls may have problems. I guess you will discover quickly.

Reply:
This defaults to `-O1`, so these backends will only have problems if given an explicit `-O0`.  However, the inlining should probably be moved to the midend as inlining being required (or not) is really target specific.

Reply:
Ideally it should be in the midend, but it does eliminate some constructs that later passes do not expect, so it may not be easy to do that.